test/mpi/threads/part/parrived_wait.c

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpi.h"
+#include <stdio.h>
+#include "mpitest.h"
+#include "mpithreadtest.h"
+
+/* Multithreaded version of test/mpi/part/parrived_wait.
+ *
+ * Each thread on the sender side prepares the partitions by using pready_range;
+ * each thread on the receiver side receives the partitions by using parrived.
+ *
+ * How to use the test:.
+ * - The send/receive partitions, total count of basic elements, and iteration of
+ *   start-test|wait periods are set by input parameters.
+ * - 4 threads are created on each side, can be changed by setting `NUM_THREADS`.
+ *
+ * Extended from Example 4.4 in the MPI 4.0 standard.*/
+
+#define NUM_THREADS 4
+
+/* Input parameters */
+static int spartitions = 0, rpartitions = 0, iteration = 0;
+static MPI_Count tot_count = 0;
+/* Internal variables */
+static double *buf = NULL;
+static MPI_Count scount = 0, rcount = 0;
+static int rank, size, thread_errs[NUM_THREADS];
+
+typedef struct {
+    int tid;
+    int iter;
+    MPI_Request req;
+} thread_arg_t;
+static thread_arg_t thread_args[NUM_THREADS];
+
+/* Utility functions */
+#define VAL(_partition, _count, _off, _iter)        \
+        (double) ((_partition) * (_count) + (_off) + (_iter))   /* Equals to `index + iter` */
```

Comment:
Suggest rid of the macro.

The reason that this macro is bad is that it introduces 4 parameters whose meaning can only be inferred from the context.

Reply:
replaced with `VAR(index, iter)` similar to previous tests.